### PR TITLE
Integrate New vm common

### PR DESCRIPTION
**Pull request by sasurobert**
Integrate New vm common

### CODE REVIEW

File: data/state/baseAccount.go [hunk 1]
```diff
@@ -37,6 +38,11 @@ func (ba *baseAccount) DataTrieTracker() DataTrieTracker {
 	return ba.dataTrieTracker
 }
 
+// AccountDataHandler return the account data handler
+func (ba *baseAccount) AccountDataHandler() vmcommon.AccountDataHandler {
```
Comment (iulianpascalau) [inline data/state/baseAccount.go:42]: Interesting pattern to avoid cyclic imports. Make 2 functions return the same thing but the function definition is different as to comply with a local interface :thinking:

File: genesis/process/shardGenesisBlockCreator.go [hunk 2]
```diff
@@ -275,11 +276,7 @@ func createProcessorsForShardGenesisBlock(arg ArgsGenesisBlockCreator, enableEpo
 		Accounts:             arg.Accounts,
 		ShardCoordinator:     arg.ShardCoordinator,
 	}
-	builtInFuncFactory, err := builtInFunctions.NewBuiltInFunctionsFactory(argsBuiltIn)
-	if err != nil {
-		return nil, err
-	}
-	builtInFuncs, err := builtInFuncFactory.CreateBuiltInFunctionContainer()
+	builtInFuncs, err := builtInFunctions.NewBuiltInFunctionsFactory(argsBuiltIn)
```
Comment (iulianpascalau) [inline genesis/process/shardGenesisBlockCreator.go:279]: I think we need a renaming here. On the old code this returned a factory that we had to call CreateBuiltInFunctionContainer. Now this returns directly the container here

File: integrationTests/vm/arwen/utils.go [hunk 3]
```diff
@@ -313,7 +312,6 @@ func (context *TestContext) initTxProcessorWithOneSCExecutorWithVMs() {
 			SetGasRefundedCalled: func(gasRefunded uint64, hash []byte) {},
 		},
 		GasSchedule:       mock.NewGasScheduleNotifierMock(gasSchedule),
-		BuiltInFunctions:  context.BlockchainHook.GetBuiltInFunctions(),
```
Comment (iulianpascalau) [inline integrationTests/vm/arwen/utils.go:316]: :+1:

File: integrationTests/vm/txsFee/txCostEstimator_test.go [hunk 4]
```diff
@@ -178,5 +178,5 @@ func TestAsyncESDTTransfer(t *testing.T) {
 
 	res, err := testContext.TxCostHandler.ComputeTransactionGasLimit(tx)
 	require.Nil(t, err)
-	require.Equal(t, uint64(99900001), res.GasUnits)
+	require.Equal(t, uint64(34207), res.GasUnits)
```
Comment (iulianpascalau) [inline integrationTests/vm/txsFee/txCostEstimator_test.go:181]: why this change?

File: process/smartContract/process.go [hunk 5]
```diff
@@ -923,26 +917,23 @@ func (sc *scProcessor) resolveBuiltInFunctions(
...
 			ReturnCode:    vmcommon.UserError,
 			ReturnMessage: err.Error(),
 			GasRemaining:  0,
 		}
+
+		return vmOutput, nil
 	}
 
-	err = sc.saveAccounts(acntSnd, acntDst)
+	acntSnd, err = sc.reloadLocalAccount(acntSnd)
```
Comment (iulianpascalau) [inline process/smartContract/process.go]: does this still works as expected?

File: process/smartContract/process.go [hunk 6]
```diff
@@ -785,11 +779,17 @@ func (sc *scProcessor) ExecuteBuiltInFunction(
...
 
 	var vmOutput *vmcommon.VMOutput
-	vmOutput, err = sc.resolveBuiltInFunctions(acntSnd, acntDst, vmInput)
+	vmOutput, err = sc.resolveBuiltInFunctions(vmInput)
 	if err != nil {
 		log.Debug("processed built in functions error", "error", err.Error())
 		return 0, err
 	}
+
+	acntSnd, err = sc.reloadLocalAccount(acntSnd)
```
Comment (iulianpascalau) [inline process/smartContract/process.go:788]: we need to reload here as to reflect the new changes in the acntSnd done in the prepareExecution, after the processSCPayment call

Comment (sasurobert) [reply process/smartContract/process.go:788]: the reload is needed to reflect the changes coming from the built in functions call.

File: genesis/process/shardGenesisBlockCreator.go (hunk 2 above)
Comment (sasurobert) [reply genesis/process/shardGenesisBlockCreator.go:279]: done

File: process/smartContract/process.go (hunk 6 above)
Comment (iulianpascalau) [reply process/smartContract/process.go:788]: No, as that function no longer calls the SaveAccount method

File: process/smartContract/builtInFunctions/factory.go [hunk 7]
```diff
@@ -21,8 +21,8 @@ type ArgsCreateBuiltInFunctionContainer struct {
 	ShardCoordinator     sharding.Coordinator
 }
 
-// NewBuiltInFunctionsFactory creates a factory which will instantiate the built in functions contracts
-func NewBuiltInFunctionsFactory(args ArgsCreateBuiltInFunctionContainer) (vmcommon.BuiltInFunctionContainer, error) {
+// CreateBuiltInFunctionContainer creates a factory which will instantiate the built in functions contracts
```
Comment (iulianpascalau) [inline process/smartContract/builtInFunctions/factory.go:24]: CreateBuiltInFunctionContainer creates a container that will hold all the available built in functions

File: data/state/baseAccount.go (hunk 1 above)
Comment (sasurobert) [reply data/state/baseAccount.go:42]: at least golang let's you do this.

**Review by iulianpascalau — DISMISSED** [review]

File: data/state/baseAccount.go (hunk 1 above)
Comment (bogdan-rosianu) [reply data/state/baseAccount.go:42]: returns*

**Review by bogdan-rosianu — CHANGES_REQUESTED** [review]

**Review by bogdan-rosianu — DISMISSED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by dorin-iancu — APPROVED** [review]
